src/thunderbots/software/ai/hl/stp/evaluation/indirect_chip.h

```diff
@@ -0,0 +1,121 @@
+     * to chipping calibration.
+     *
+     * @param World Object
+     *
+     * @return Point to chip and chase at; If empty, point is not valid
+     */
+    std::optional<Point> indirect_chip_and_chase_target(const World& world);
+
+    /**
+     * Returns the target point that the chipper will shoot at and the chaser will meet
+     * ball at. Given the filtered vector of triangles, determines if the triangles are
+     * empty and if the largest triangle is within reach. The target is where ball will
+     * land according to chipping calibration.
+     *
+     * @param triangles A vector of triangles that is already filtered
+     * @param ball_position Position of the ball
+     *
+     * @return Point to chip and chase at; If empty, point is not valid
+     */
+    std::optional<Point> indirect_chip_and_chase_target(
+        const std::vector<Triangle>& triangles, Point ball_position);
+
+    /**
+     * Creates a vector of triangles that picks all permutations of points from the
+     * list of all non-goalie enemy players as well as the four points returned by
+     * get_chip_area_target_corners.
+     *
+     * @param World Object
+     * @param enemy_players
+     *
+     * @return Vector of triangles
+     */
+    std::vector<Triangle> get_all_triangles(const World& world,
+                                            std::vector<Point> enemy_players);
+
+    /**
+     * Given a vector of triangles, returns a new vector that only contains triangles
+     * without enemy robots.
+     *
+     * @param triangles
+     * @param enemy_players
+     *
+     * @return Vector of triangles with no enemy robots
+     */
+    std::vector<Triangle> filter_open_triangles(std::vector<Triangle> triangles,
+                                                std::vector<Point> enemy_players);
+
+    /**
+     * Returns the center point and area of the given triangle.
+     *
+     * @param triangle
+     *
+     * @return Centre point of triangle
+     * @return Area of triangle
+     */
+    std::pair<Point, double> get_triangle_center_and_area(Triangle triangle);
+
+    /**
+     * Remove all Triangles in a given list whose centers do not fall
+     * within the rectangle returned by get_chip_target_area.
+     *
+     * @param World Object
+     * @param triangles
+     *
+     * @return Valid triangles
+     */
+    std::vector<Triangle> remove_outofbounds_triangles(const World& world,
+                                                       std::vector<Triangle> triangles);
+
+    /**
+     * Returns four Points representing a rectangular region to chip and
+     * chase. This rectangle is from ball's position to enemy's end of field.
+     * All points are 'inset' distance away from each edge of the field to allow a
+     * buffer for catching and prevent ball from leaving the field.
+     *
+     * @param World Object
+     * @param inset Distance away from each edge of field
+     *
+     * @return Four points for rectangle
```

Comment:
Perhaps we could return a `Rectangle` instead? I believe our `Rectangle` class is setup so you can do the following:
```
Rectangle r(Point(0,0), Point(5,10));
Point p1 = r[0];
Point p2 = r[1];
...
```
(this is specified on line 97 of the `rectangle.h` file)

Reply:
I agree with returning a `Rectangle` instead!
How would you construct a `Rectangle` if you're not 100% sure about which two points are diagonally opposite from each other?

I'm not too sure what the thing on line 97 does (I really do see the importance of easy-to-understand documentation!) In your comment, 
```
Rectangle r(Point(0,0), Point(5,10));
```
Are `Point(0,0)` and `Point(5,10)` two pre-determined, diagonally opposite points?

Do we use the returned `"digit corresponding to a specific corner"` (I assume that's `r[0]`, `r[1]`...) to determine which corners are diagonally opposite?

Reply:
Ah, so if you take a look at `geom/rectangle.h`, it explains what that constructor does (your guess was right, they are the diagonally opposite points!). 

In this case I think we can be certain that `Point(ballX, negFieldY)` and `Point(fieldX, posFieldY)` are diagonally opposite? My logic here is that we've changed both x and y, and if you have the four corners of a rectangle, the any two corners with both different x and different y values will be diagonally opposite.

Reply:
After some fidgeting, here's what I got:
```
Rectangle target_rectangle = Rectangle(Point(ballX, negFieldY), Point(fieldX, posFieldY));
Point p1 = r[0]; // south-west corner
Point p2 = r[1]; // north-west corner
Point p3 = r[2]; // north-east corner
Point p4 = r[3]; // south-east corner

return target_rectangle;
```

Since `findBestChipTargetArea` will now return a `Rectangle` instead of `std::vector<Point>`, implementation changes will be made in other functions that uses this (WIP).

I see your logic about the diagonal corners. This is my current interpretation:
![chip target area corners](https://user-images.githubusercontent.com/38922275/54338671-c409fd00-45ef-11e9-9709-20cc127decf9.png)